### PR TITLE
Add 'Bucket.list_notifications' API wrapper.

### PR DESCRIPTION
**Pull request by tseaver**
Uses #3983 as a base.

Toward #3956.

### CODE REVIEW

File: storage/google/cloud/storage/bucket.py [hunk 1]
```diff
@@ -405,6 +424,30 @@ def list_blobs(self, max_results=None, page_token=None, prefix=None,
         iterator.prefixes = set()
         return iterator
 
+    def list_notifications(self, client=None):
+        """List pubsub notifications for this bucket.
```
Comment (dhermes) [inline storage/google/cloud/storage/bucket.py]: "Pub / Sub"

File: storage/google/cloud/storage/notification.py [hunk 2]
```diff
@@ -82,6 +90,40 @@ def __init__(self, bucket, topic_name,
...
+        :param resource: JSON repr of the notification
+
+        :type bucket: :class:`google.cloud.storage.bucket.Bucket`
+        :param bucket: Bucket to which the notification is bound.
+
+        :rtype: :class:`BucketNotification`
+        :returns: the new notification instance
+        :raises ValueError:
+            if resource is missing 'topic' key, or if it is not formatted
+            per documented spec.
```
Comment (dhermes) [inline storage/google/cloud/storage/notification.py]: We should help our users here and explain what our interpretation of "document spec" is.

File: storage/google/cloud/storage/notification.py [hunk 3]
```diff
@@ -82,6 +90,40 @@ def __init__(self, bucket, topic_name,
...
+        """
+        topic_path = resource.get('topic')
+        if topic_path is None:
+            raise ValueError('Resource has no topic')
+
+        match = _TOPIC_REF_RE.match(topic_path)
+        if match is None:
+            raise ValueError(
+                'Resource has invalid topic: {}'.format(topic_path))
```
Comment (dhermes) [inline storage/google/cloud/storage/notification.py]: You could include the regex in this error message? (I assume it has a not-so-bad repr?)

File: storage/google/cloud/storage/bucket.py (hunk 1 above)
Comment (tseaver) [reply storage/google/cloud/storage/bucket.py]: 998b9f9

File: storage/google/cloud/storage/notification.py (hunk 2 above)
Comment (tseaver) [reply storage/google/cloud/storage/notification.py]: 998b9f9

File: storage/google/cloud/storage/notification.py (hunk 3 above)
Comment (tseaver) [reply storage/google/cloud/storage/notification.py]: 998b9f9 puts the URL of the spec in the message.

**Comment by tseaver** [comment]
@dhermes Any remaining issues?

**Pull request merged**